compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+
+#define DEBUG_TYPE "global-loop-invariant-code-motion"
+#define LICM_DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
+
+// Check if the op is a leaf op we always want to hoist.
+static bool isHoistableLeafOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return isa<tensor::PackOp>(op);
+}
+
+// Check if the op is hoistable (but we might not want to hoist it alone).
+static bool isHoistableOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return op->hasTrait<OpTrait::ConstantLike>() || isa<tensor::EmptyOp>(op) ||
+         isHoistableLeafOp(op);
+}
+
+// Check if the op and its producers are loop invariants and hoistable. Results
+// are cached in hoistableOpMap to avoid repeated traversals.
+static bool checkHoistableBackwardSlice(
+    LoopLikeOpInterface loopOp, Operation *op,
+    llvm::SmallDenseMap<Operation *, bool> &hoistableOpMap) {
+  // First check if the op has been analyzed.
+  if (hoistableOpMap.contains(op))
+    return hoistableOpMap[op];
+
+  bool hoistable = true;
+  // Currently only hoist ops with no region (so no implicit capture).
+  if (op->getNumRegions() > 0) {
+    hoistable = false;
+  } else {
+    // Check if all producers are hoistable.
+    for (OpOperand &operand : op->getOpOperands()) {
+      Value value = operand.get();
+      // Ignore values defined outside the loop.
+      if (loopOp.isDefinedOutsideOfLoop(value))
+        continue;
+
+      Operation *producer = value.getDefiningOp();
+      // If the value is not an operation, we don't hoist it.
+      if (!producer ||
+          !checkHoistableBackwardSlice(loopOp, producer, hoistableOpMap)) {
```

Comment:
Instead of recursing can we just use a work list here

Reply:
I was thinking about this but currently I still think it will be more difficult to implement with worklist.

The reason is that we want to know if the root op's operands can be hoisted together, so the same check also needs to be done on the operands, and so on, recursively. We only know the root op is hoistable after all check results returned from its operands, so the recursive works well here.

In my mind to get rid of recursive, we need a stack instead of worklist, means that the root op will push its operands into the stack and stay in the bottom until all its operands are checked. I can refactor with a stack but I feel like recursive is easier to understand.